File: day03/01_1.rb
file = File.open('input.txt', 'r')

arr = []

file.each_line do |line|
  aa, a = line.split('@')
  pos, size = a.split(':')
  x, y = pos.split(',').map(&:to_i)
  lenX, lenY = size.split('x').map(&:to_i)
  arr.push({
    x1: x + 1,
    x2: x + lenX,
    y1: y + 1,
    y2: y + lenY,
  })
end
file.close

$mem = {}

def intersect_area(rec1, rec2)
  area = 0
  (rec1[:x1]..rec1[:x2]).each do |x|
    (rec1[:y1]..rec1[:y2]).each do |y|
      if (rec2[:x1] <= x and x <= rec2[:x2] and rec2[:y1] <= y and y <= rec2[:y2]) then
        if (!$mem.has_key?("#{x},#{y}")) then
          area += 1
        end
        $mem["#{x},#{y}"] = true
      end
    end
  end
  area
end

ans = 0

(0...arr.size).each do |i|
  (i+1...arr.size).each do |j|
    ans += intersect_area(arr[i], arr[j])
  end
end

puts "Ans: #{ans}"

File: day03/01_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type position struct {
	x1 int
	x2 int
	y1 int
	y2 int
}

func toMap(pos position) map[string]int {
	m := make(map[string]int)
	for x := pos.x1; x <= pos.x2; x++ {
		for y := pos.y1; y <= pos.y2; y++ {
			k := strconv.Itoa(x) + "," + strconv.Itoa(y)
			m[k] = 1
		}
	}
	return m
}

func divide(list []position) map[string]int {
	if len(list) == 0 {
		return make(map[string]int)
	}

	if len(list) == 1 {
		return toMap(list[0])
	}

	halfSize := int(len(list) / 2)
	listA := divide(list[:halfSize])
	listB := divide(list[halfSize:])

	mergedMap := map[string]int{}

	for k, v := range listA {
		mergedMap[k] = v
	}

	for k, v := range listB {
		if vv, ok := mergedMap[k]; ok {
			mergedMap[k] = vv + v
		} else {
			mergedMap[k] = v
		}
	}

	return mergedMap
}

func main() {
	file, _ := os.Open("input.txt")
	// file, _ := os.Open("sample.txt")

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var input []string

	var arr []position

	for scanner.Scan() {
		input = append(input, scanner.Text())
	}

	for i := 0; i < len(input); i++ {
		r := strings.Split(input[i], "@ ")
		ps := strings.Split(r[1], ": ")
		xy := strings.Split(ps[0], ",")
		len := strings.Split(ps[1], "x")
		x, _ := strconv.Atoi(xy[0])
		y, _ := strconv.Atoi(xy[1])
		lenX, _ := strconv.Atoi(len[0])
		lenY, _ := strconv.Atoi(len[1])

		o := position{
			x1: (x + 1),
			x2: (x + lenX),
			y1: (y + 1),
			y2: (y + lenY),
		}

		arr = append(arr, o)
	}

	positionMap := divide(arr)

	ans := 0
	for _, v := range positionMap {
		if v >= 2 {
			ans++
		}
	}

	fmt.Println(ans)
}
